code/controllers/subsystem/garbage.dm

```diff
@@ -97,11 +103,24 @@ SUBSYSTEM_DEF(garbage)
 				queue = GC_QUEUE_CHECK+1
 			if (GC_QUEUE_HARDDELETE)
 				HandleQueue(GC_QUEUE_HARDDELETE)
-				if (state == SS_PAUSED) //make us wait again before the next run.
-					state = SS_RUNNING
 				break
 
-	
+	if (state == SS_PAUSED) //make us wait again before the next run.
+		state = SS_RUNNING
+
+/datum/controller/subsystem/garbage/proc/writeMemorial()
+	to_chat(world, "<span class='userdanger'>A diagnostics file is being collected to aid debugging a likely impending OOM crash. The server will be unresponsive for around a minute.</span>")
+	sleep(5) // To make sure the message is seen
+	var/list/results = new /list(GC_QUEUE_COUNT)
+	results[GC_QUEUE_PREQUEUE] = list()
+	for (var/item in queues[GC_QUEUE_PREQUEUE]) // Prequeue items are references
+		results[GC_QUEUE_PREQUEUE][item?:type] += 1
+	for(var/idx in 2 to GC_QUEUE_COUNT) // The rest are not
+		results[idx] = list()
+		for (var/ref in queues[idx])
+			var/item = locate(ref)
+			results[idx][item?:type] += 1
```

Comment:
if (item?:gc_destroyed == queues[ref])
  results[idx][item?:type] += 1

This is logging ref reuse (new datum reuses the \ref of a deleted datum still in the queue) because its not checking the marker we placed on the datum to prevent such cases.